File: apps/viewer/src/physics/solver.ts
/**
 * Over-damped implicit rod solver - TypeScript port of
 * `packages/physics-core/src/cathsim_physics/solver.py`.
 *
 * Time integration is the minimising-movement (backward-Euler on over-damped
 * dynamics) scheme
 *
 *     q^{n+1} = argmin_q [ E(q) + 1/(2 dt) sum_k c_k (q_k - q_k^n)^2 ]
 *
 * which is unconditionally stable - there is no CFL limit coming from the very
 * stiff bending term - and reduces to static equilibrium as dt -> infinity.
 *
 * The minimisation is limited-memory BFGS with a backtracking Armijo line
 * search.  A rod couples a stiff local stretch mode to a soft global bending
 * mode, so the history size matters a great deal: the Python reference needs
 * 14624 iterations with 20 corrections and 607 with 150.  The default here is
 * chosen accordingly.
 *
 * Research prototype - Not for clinical use.
 */

import {
  type EnergyBreakdown,
  type ExternalLoads,
  type RodMaterial,
  type RodRest,
  type RodState,
  RodGeometryError,
  energyAndGradient,
  nodeCount,
} from './rod';

export interface BoundaryConditions {
  /** Node indices whose position is prescribed (proximal clamp). */
  clampedNodes: number[];
  /** Edge indices whose twist angle is prescribed (handle roll). */
  clampedEdges: number[];
}

export interface SolverSettings {
  maxIterations: number;
  /** Convergence threshold on the infinity norm of the free-DOF gradient [N]. */
  gradientToleranceN: number;
  historySize: number;
  /**
   * Wall-clock budget for one solve [ms]; 0 disables the limit.
   *
   * A rod couples a soft global bending mode (tip stiffness ~ 3EI/L^3) to a
   * stiff local stretch mode (~EA/h), a condition number around 1e5, so a
   * *converged* implicit step can need several hundred L-BFGS iterations.
   * Interactive frames cannot always afford that.  The budget keeps the frame
   * rate honest, and the shortfall is never hidden: `converged` goes false and
   * the residual is displayed.  The rod still descends the same physical
   * energy, so it keeps approaching the correct equilibrium over subsequent
   * frames - and `solveStatic` always reaches it exactly.
   */
  maxWallTimeMs: number;
}

export interface SolverReport {
  iterations: number;
  residual: number;
  maxPenetrationMm: number;
  converged: boolean;
  energy: EnergyBreakdown;
  /** Wall-clock duration of the solve [ms]. */
  wallTimeMs: number;
}

export const DEFAULT_SETTINGS: SolverSettings = {
  maxIterations: 600,
  gradientToleranceN: 1e-6,
  historySize: 60,
  maxWallTimeMs: 12,
};

/** Reduced-cost preset offered in the UI when the frame budget is missed. */
export const LOW_QUALITY_SETTINGS: SolverSettings = {
  maxIterations: 150,
  gradientToleranceN: 1e-5,
  historySize: 30,
  maxWallTimeMs: 6,
};

export class RodSolverError extends Error {}

/**
 * Solves one rod. The instance owns scratch buffers so that stepping does not
 * allocate per frame (see the performance targets in the project contract).
 */
export class RodSolver {
  private readonly freeIndices: Int32Array;

  private readonly gradNodes: Float64Array;

  private readonly gradPhi: Float64Array;

  private readonly nodeDamping: Float64Array;

  private readonly edgeDamping: Float64Array;

  private readonly trial: RodState;

  private readonly previousNodes: Float64Array;

  private readonly previousPhi: Float64Array;

  private history: { s: Float64Array; y: Float64Array; rho: number }[] = [];

  constructor(
    private readonly material: RodMaterial,
    private readonly rest: RodRest,
    template: RodState,
    boundary: BoundaryConditions,
    public settings: SolverSettings = DEFAULT_SETTINGS,
  ) {
    const n = nodeCount(template);
    const edges = n - 1;
    const clampedNode = new Set(boundary.clampedNodes);
    const clampedEdge = new Set(boundary.clampedEdges);

    const free: number[] = [];
    for (let i = 0; i < n; i += 1) {
      if (!clampedNode.has(i)) free.push(3 * i, 3 * i + 1, 3 * i + 2);
    }
    // Twist DOFs are appended after the positional ones, offset by 3N.
    for (let j = 0; j < edges; j += 1) {
      if (!clampedEdge.has(j)) free.push(3 * n + j);
    }
    if (free.length === 0) {
      throw new RodSolverError('Every degree of freedom is clamped; there is nothing to solve.');
    }
    this.freeIndices = Int32Array.from(free);
    this.gradNodes = new Float64Array(3 * n);
    this.gradPhi = new Float64Array(edges);
    this.previousNodes = new Float64Array(3 * n);
    this.previousPhi = new Float64Array(edges);
    this.trial = {
      nodesMm: new Float64Array(3 * n),
      phiRad: new Float64Array(edges),
      d1First: template.d1First,
    };

    // Damping weights: per-unit-length coefficients lumped onto nodes/edges.
    this.nodeDamping = new Float64Array(n);
    for (let j = 0; j < edges; j += 1) {
      const half = 0.5 * rest.edgeLengthsMm[j];
      this.nodeDamping[j] += material.dampingNsPerMm2 * half;
      this.nodeDamping[j + 1] += material.dampingNsPerMm2 * half;
    }
    this.edgeDamping = new Float64Array(edges);
    for (let j = 0; j < edges; j += 1) {
      this.edgeDamping[j] = material.torsionalDampingNmmSPerRad * rest.edgeLengthsMm[j];
    }
  }

  /** Drop the limited-memory curvature history (after a reset or a re-mesh). */
  resetHistory(): void {
    this.history = [];
  }

  /** Relax to static equilibrium (no inertia term). */
  solveStatic(state: RodState, loads?: ExternalLoads): SolverReport {
    this.resetHistory();
    return this.minimise(state, loads, 0);
  }

  /** Advance the over-damped dynamics by `dtS` seconds. */
  step(state: RodState, dtS: number, loads?: ExternalLoads): SolverReport {
    if (!Number.isFinite(dtS) || dtS <= 0) {
      throw new RodSolverError(
        `Time step must be a positive finite number of seconds, got ${String(dtS)}.`,
      );
    }
    this.previousNodes.set(state.nodesMm);
    this.previousPhi.set(state.phiRad);
    return this.minimise(state, loads, 1 / dtS);
  }

  /** Free-DOF gradient assembled from the last energy evaluation. */
  private packGradient(out: Float64Array, nodeOffset: number): void {
    for (let k = 0; k < this.freeIndices.length; k += 1) {
      const index = this.freeIndices[k];
      out[k] = index < nodeOffset ? this.gradNodes[index] : this.gradPhi[index - nodeOffset];
    }
  }

  private objective(
    candidate: Float64Array,
    state: RodState,
    loads: ExternalLoads | undefined,
    dampingScalePerS: number,
    gradientOut: Float64Array,
  ): number {
    const nodeOffset = state.nodesMm.length;
    this.trial.nodesMm.set(state.nodesMm);
    this.trial.phiRad.set(state.phiRad);
    for (let k = 0; k < this.freeIndices.length; k += 1) {
      const index = this.freeIndices[k];
      if (index < nodeOffset) this.trial.nodesMm[index] = candidate[k];
      else this.trial.phiRad[index - nodeOffset] = candidate[k];
    }

    const energy = energyAndGradient(
      this.trial,
      this.rest,
      this.material,
      loads,
      this.gradNodes,
      this.gradPhi,
    );
    let total = energy.total;

    if (dampingScalePerS > 0) {
      for (let i = 0; i < this.nodeDamping.length; i += 1) {
        const weight = dampingScalePerS * this.nodeDamping[i];
        for (let k = 0; k < 3; k += 1) {
          const delta = this.trial.nodesMm[3 * i + k] - this.previousNodes[3 * i + k];
          total += 0.5 * weight * delta * delta;
          this.gradNodes[3 * i + k] += weight * delta;
        }
      }
      for (let j = 0; j < this.edgeDamping.length; j += 1) {
        const weight = dampingScalePerS * this.edgeDamping[j];
        const delta = this.trial.phiRad[j] - this.previousPhi[j];
        total += 0.5 * weight * delta * delta;
        this.gradPhi[j] += weight * delta;
      }
    }

    this.packGradient(gradientOut, nodeOffset);
    return total;
  }

  /** L-BFGS with a backtracking Armijo line search. */
  private minimise(
    state: RodState,
    loads: ExternalLoads | undefined,
    dampingScalePerS: number,
  ): SolverReport {
    const startMs = now();
    const dimension = this.freeIndices.length;
    const nodeOffset = state.nodesMm.length;

    const x = new Float64Array(dimension);
    for (let k = 0; k < dimension; k += 1) {
      const index = this.freeIndices[k];
      x[k] = index < nodeOffset ? state.nodesMm[index] : state.phiRad[index - nodeOffset];
    }

    const gradient = new Float64Array(dimension);
    const direction = new Float64Array(dimension);
    const candidate = new Float64Array(dimension);
    const nextGradient = new Float64Array(dimension);

    // The curvature history is kept ACROSS calls.  With a heavily over-damped
    // rod the implicit step is close to a full static solve, so a frame-budgeted
    // solve alone would stall: every frame would spend its whole budget
    // re-discovering the same curvature.  Carrying the history over turns the
    // per-frame budget into a continuation solve that keeps descending the same
    // physical energy and converges to the true equilibrium over a few frames.
    // It is discarded whenever the search direction stops being a descent
    // direction (see below) or on `resetHistory()`.
    const history = this.history;
    const alpha = new Float64Array(this.settings.historySize);

    let value: number;
    try {
      value = this.objective(x, state, loads, dampingScalePerS, gradient);
    } catch (error) {
      throw this.wrap(error);
    }

    let iterations = 0;
    let residual = infinityNorm(gradient);

    const deadlineMs =
      this.settings.maxWallTimeMs > 0 ? startMs + this.settings.maxWallTimeMs : Infinity;

    for (; iterations < this.settings.maxIterations; iterations += 1) {
      if (residual <= this.settings.gradientToleranceN) break;
      if (iterations % 8 === 0 && now() > deadlineMs) break;

      // --- two-loop recursion
      direction.set(gradient);
      for (let h = history.length - 1; h >= 0; h -= 1) {
        const { s, y, rho } = history[h];
        alpha[h] = rho * dot(s, direction);
        axpy(direction, y, -alpha[h]);
      }
      if (history.length > 0) {
        const last = history[history.length - 1];
        const scale = dot(last.s, last.y) / dot(last.y, last.y);
        if (Number.isFinite(scale) && scale > 0) {
          for (let k = 0; k < dimension; k += 1) direction[k] *= scale;
        }
      }
      for (let h = 0; h < history.length; h += 1) {
        const { s, y, rho } = history[h];
        const beta = rho * dot(y, direction);
        axpy(direction, s, alpha[h] - beta);
      }
      for (let k = 0; k < dimension; k += 1) direction[k] = -direction[k];

      let slope = dot(gradient, direction);
      if (!(slope < 0)) {
        // The approximated inverse Hessian lost positive definiteness; restart.
        for (let k = 0; k < dimension; k += 1) direction[k] = -gradient[k];
        slope = -dot(gradient, gradient);
        history.length = 0;
      }

      // --- backtracking Armijo line search
      let stepSize = history.length === 0 ? Math.min(1, 1 / Math.max(residual, 1e-12)) : 1;
      let nextValue = value;
      let accepted = false;
      for (let attempt = 0; attempt < 40; attempt += 1) {
        for (let k = 0; k < dimension; k += 1) candidate[k] = x[k] + stepSize * direction[k];
        try {
          nextValue = this.objective(candidate, state, loads, dampingScalePerS, nextGradient);
        } catch (error) {
          if (error instanceof RodGeometryError) {
            // The trial configuration folded; shorten the step instead of failing.
            stepSize *= 0.5;
            continue;
          }
          throw this.wrap(error);
        }
        if (nextValue <= value + 1e-4 * stepSize * slope) {
          accepted = true;
          break;
        }
        stepSize *= 0.5;
      }
      if (!accepted) break;

      const s = new Float64Array(dimension);
      const y = new Float64Array(dimension);
      for (let k = 0; k < dimension; k += 1) {
        s[k] = candidate[k] - x[k];
        y[k] = nextGradient[k] - gradient[k];
      }
      const sy = dot(s, y);
      if (sy > 1e-16) {
        history.push({ s, y, rho: 1 / sy });
        if (history.length > this.settings.historySize) history.shift();
      }

      x.set(candidate);
      gradient.set(nextGradient);
      value = nextValue;
      residual = infinityNorm(gradient);
    }

    // Write the solution back into the caller's state.
    for (let k = 0; k < dimension; k += 1) {
      const index = this.freeIndices[k];
      if (index < nodeOffset) state.nodesMm[index] = x[k];
      else state.phiRad[index - nodeOffset] = x[k];
    }

    const finite = state.nodesMm.every(Number.isFinite) && state.phiRad.every(Number.isFinite);
    if (!finite) {
      throw new RodSolverError(
        'The solver produced a non-finite configuration and was stopped. Reduce ' +
          'the time step or check the parameter profile.',
      );
    }
    // Report the elastic energy of the *accepted* solution, not of whatever
    // trial point the line search happened to evaluate last.
    const energy = energyAndGradient(
      state,
      this.rest,
      this.material,
      loads,
      this.gradNodes,
      this.gradPhi,
    );
    return {
      iterations,
      residual,
      maxPenetrationMm: 0, // populated by the contact stage (Milestone 2)
      converged: residual <= this.settings.gradientToleranceN,
      energy,
      wallTimeMs: now() - startMs,
    };
  }

  private wrap(error: unknown): RodSolverError {
    return error instanceof Error
      ? new RodSolverError(error.message)
      : new RodSolverError(String(error));
  }
}

function now(): number {
  return typeof performance !== 'undefined' ? performance.now() : Date.now();
}

function dot(a: Float64Array, b: Float64Array): number {
  let sum = 0;
  for (let i = 0; i < a.length; i += 1) sum += a[i] * b[i];
  return sum;
}

function axpy(target: Float64Array, source: Float64Array, factor: number): void {
  for (let i = 0; i < target.length; i += 1) target[i] += factor * source[i];
}

function infinityNorm(a: Float64Array): number {
  let max = 0;
  for (let i = 0; i < a.length; i += 1) {
    const value = Math.abs(a[i]);
    if (value > max) max = value;
  }
  return max;
}

File: packages/physics-core/src/cathsim_physics/solver.py
"""Quasi-static / over-damped solver for the discrete elastic rod.

Time integration
----------------
Catheter manipulation at clinical speeds is heavily damped and essentially
quasi-static.  We therefore integrate the rod with the *minimising movement*
(implicit gradient-flow / backward-Euler on over-damped dynamics) scheme

.. code-block:: text

    q^{n+1} = argmin_q [ E(q) + 1/(2 dt) sum_k  c_k (q_k - q_k^n)^2 ]

whose stationarity condition is  ``c (q^{n+1} - q^n)/dt = -grad E(q^{n+1})``,
i.e. backward Euler applied to ``c q_dot = -grad E``.  The scheme is
unconditionally stable (no CFL limit from the stiff bending term), is
energy-dissipating by construction, and reduces to the static equilibrium
problem as ``dt -> inf``.  That gives us the two solver entry points below:
:meth:`RodSolver.solve_static` and :meth:`RodSolver.step`.

Both are driven by L-BFGS-B on the *free* degrees of freedom with the exact
analytic gradient from :mod:`cathsim_physics.rod`.

Research prototype - Not for clinical use.
"""

from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np
from scipy.optimize import minimize

from .rod import (
    EnergyBreakdown,
    ExternalLoads,
    RodMaterial,
    RodRest,
    RodState,
    energy_and_gradient,
)


@dataclass(frozen=True)
class BoundaryConditions:
    """Which degrees of freedom are prescribed rather than solved for.

    A proximally clamped catheter fixes node 0 and node 1 (position + tangent)
    and the twist angle of edge 0 (the handle roll).  Everything distal is free.

    Attributes:
        clamped_nodes: indices of nodes whose position is prescribed.
        clamped_edges: indices of edges whose twist angle is prescribed.
    """

    clamped_nodes: tuple[int, ...] = (0, 1)
    clamped_edges: tuple[int, ...] = (0,)


@dataclass(frozen=True)
class SolverSettings:
    """Numerical settings of the solver.

    Attributes:
        max_iterations: L-BFGS-B iteration cap per call.
        gradient_tolerance_n: convergence threshold on the infinity norm of the
            free-DOF gradient.  Positional entries are newtons, twist entries
            newton-millimetres; the same scalar threshold is used for both,
            which is why it is reported next to the residual in the UI.
        history_size: L-BFGS-B memory.  A rod couples a very stiff local
            stretch mode to a very soft global bending mode (condition number
            ~1e5 for a 80 mm shaft), so a *large* limited-memory history pays
            for itself: on the demo steering case, 20 corrections need 14624
            iterations while 150 corrections need 607.  See
            ``docs/physics-model.md``.
    """

    max_iterations: int = 3000
    gradient_tolerance_n: float = 1.0e-6
    history_size: int = 150


@dataclass
class SolverReport:
    """Diagnostics of a single solver call, mirrored into ``SimulationFrame``."""

    iterations: int = 0
    residual: float = 0.0
    max_penetration_mm: float = 0.0
    converged: bool = True
    energy: EnergyBreakdown = field(default_factory=EnergyBreakdown)
    message: str = ""


class RodSolverError(RuntimeError):
    """Raised when the solver cannot produce a usable configuration."""


class RodSolver:
    """Over-damped implicit solver for a single :class:`RodState`."""

    def __init__(
        self,
        material: RodMaterial,
        rest: RodRest,
        boundary: BoundaryConditions | None = None,
        settings: SolverSettings | None = None,
    ) -> None:
        self.material = material
        self.rest = rest
        self.boundary = boundary or BoundaryConditions()
        self.settings = settings or SolverSettings()

    # -- degree-of-freedom packing ---------------------------------------

    def _free_masks(self, state: RodState) -> tuple[np.ndarray, np.ndarray]:
        node_mask = np.ones(state.n_nodes, dtype=bool)
        node_mask[list(self.boundary.clamped_nodes)] = False
        edge_mask = np.ones(state.n_edges, dtype=bool)
        edge_mask[list(self.boundary.clamped_edges)] = False
        return node_mask, edge_mask

    def _pack(self, state: RodState, node_mask: np.ndarray, edge_mask: np.ndarray) -> np.ndarray:
        return np.concatenate([state.nodes_mm[node_mask].ravel(), state.phi_rad[edge_mask]])

    def _unpack(
        self,
        vector: np.ndarray,
        template: RodState,
        node_mask: np.ndarray,
        edge_mask: np.ndarray,
    ) -> RodState:
        n_free_nodes = int(node_mask.sum())
        nodes = template.nodes_mm.copy()
        phi = template.phi_rad.copy()
        nodes[node_mask] = vector[: 3 * n_free_nodes].reshape(n_free_nodes, 3)
        phi[edge_mask] = vector[3 * n_free_nodes :]
        return RodState(nodes, phi, template.d1_first)

    # -- solves ----------------------------------------------------------

    def solve_static(
        self,
        state: RodState,
        loads: ExternalLoads | None = None,
    ) -> tuple[RodState, SolverReport]:
        """Relax to static equilibrium (``dt -> inf``, no inertia term)."""
        return self._minimise(state, loads, previous=None, damping_scale_per_s=0.0)

    def step(
        self,
        state: RodState,
        dt_s: float,
        loads: ExternalLoads | None = None,
    ) -> tuple[RodState, SolverReport]:
        """Advance the over-damped dynamics by ``dt_s`` seconds."""
        if not np.isfinite(dt_s) or dt_s <= 0.0:
            raise RodSolverError(
                f"Time step must be a positive finite number of seconds, got {dt_s!r}."
            )
        return self._minimise(state, loads, previous=state, damping_scale_per_s=1.0 / dt_s)

    def _dissipation_weights(self, state: RodState) -> tuple[np.ndarray, np.ndarray]:
        """Per-node [N s/mm] and per-edge [N mm s/rad] damping weights."""
        half = 0.5 * self.rest.edge_lengths_mm
        node_length = np.zeros(state.n_nodes)
        node_length[:-1] += half
        node_length[1:] += half
        node_damping = self.material.damping_n_s_per_mm * node_length
        edge_damping = self.material.torsional_damping_n_mm_s_per_rad * self.rest.edge_lengths_mm
        return node_damping, edge_damping

    def _minimise(
        self,
        state: RodState,
        loads: ExternalLoads | None,
        previous: RodState | None,
        damping_scale_per_s: float,
    ) -> tuple[RodState, SolverReport]:
        node_mask, edge_mask = self._free_masks(state)
        if not node_mask.any():
            raise RodSolverError("Every node is clamped; there is nothing to solve.")

        node_damping, edge_damping = self._dissipation_weights(state)
        prev_nodes = previous.nodes_mm.copy() if previous is not None else None
        prev_phi = previous.phi_rad.copy() if previous is not None else None

        def objective(vector: np.ndarray) -> tuple[float, np.ndarray]:
            trial = self._unpack(vector, state, node_mask, edge_mask)
            energy, grad_nodes, grad_phi = energy_and_gradient(
                trial, self.rest, self.material, loads
            )
            total = energy.total
            if prev_nodes is not None and damping_scale_per_s > 0.0:
                dx = trial.nodes_mm - prev_nodes
                dphi = trial.phi_rad - prev_phi
                total += (
                    0.5
                    * damping_scale_per_s
                    * float(np.sum(node_damping[:, None] * dx**2) + np.sum(edge_damping * dphi**2))
                )
                grad_nodes = grad_nodes + damping_scale_per_s * node_damping[:, None] * dx
                grad_phi = grad_phi + damping_scale_per_s * edge_damping * dphi
            gradient = np.concatenate([grad_nodes[node_mask].ravel(), grad_phi[edge_mask]])
            return total, gradient

        x0 = self._pack(state, node_mask, edge_mask)
        try:
            result = minimize(
                objective,
                x0,
                jac=True,
                method="L-BFGS-B",
                options={
                    "maxiter": self.settings.max_iterations,
                    "maxcor": self.settings.history_size,
                    "gtol": self.settings.gradient_tolerance_n,
                    "ftol": 1.0e-16,
                },
            )
        except ValueError as exc:  # raised by rod.py on NaN / folded geometry
            raise RodSolverError(str(exc)) from exc

        solved = self._unpack(result.x, state, node_mask, edge_mask)
        energy, grad_nodes, grad_phi = energy_and_gradient(solved, self.rest, self.material, loads)
        residual = float(
            max(
                np.abs(grad_nodes[node_mask]).max(initial=0.0),
                np.abs(grad_phi[edge_mask]).max(initial=0.0),
            )
        )
        report = SolverReport(
            iterations=int(result.nit),
            residual=residual,
            max_penetration_mm=0.0,  # populated by the contact stage (Milestone 2)
            converged=bool(np.all(np.isfinite(solved.nodes_mm)))
            and (bool(result.success) or residual <= self.settings.gradient_tolerance_n),
            energy=energy,
            message=str(result.message),
        )
        return solved, report
